packages/core/src/DirectoryLoader.ts

```diff
@@ -0,0 +1,339 @@
+			...(resources && { resources }),
+			...(alias && { alias }),
+		};
+	}
+
+	/**
+	 * Adds a node codex `categories` and `subcategories` (if defined)
+	 * to a node description `codex` property.
+	 */
+	private addCodex({
+		node,
+		filePath,
+		isCustom,
+	}: {
+		node: INodeType | IVersionedNodeType;
+		filePath: string;
+		isCustom: boolean;
+	}) {
+		try {
+			const codex = this.getCodex(filePath);
+
+			if (isCustom) {
+				codex.categories = codex.categories
+					? codex.categories.concat(CUSTOM_NODES_CATEGORY)
+					: [CUSTOM_NODES_CATEGORY];
+			}
+
+			node.description.codex = codex;
+		} catch (_) {
+			Logger.debug(`No codex available for: ${filePath.split('/').pop() ?? ''}`);
+
+			if (isCustom) {
+				node.description.codex = {
+					categories: [CUSTOM_NODES_CATEGORY],
+				};
+			}
+		}
+	}
+
+	private fixIconPath(
+		obj: INodeTypeDescription | INodeTypeBaseDescription | ICredentialType,
+		filePath: string,
+	) {
+		if (obj.icon?.startsWith('file:')) {
+			const iconPath = path.join(path.dirname(filePath), obj.icon.substring(5));
+			const relativePath = path.relative(this.directory, iconPath);
+			obj.icon = `file:${relativePath}`;
+		}
+	}
+}
+
+/**
+ * Loader for source files of nodes and credentials located in a custom dir,
+ * e.g. `~/.n8n/custom`
+ */
+export class CustomDirectoryLoader extends DirectoryLoader {
+	override async loadAll() {
+		const filePaths = await glob('**/*.@(node|credentials).js', {
+			cwd: this.directory,
+			absolute: true,
+		});
+
+		for (const filePath of filePaths) {
+			const [fileName, type] = path.parse(filePath).name.split('.');
+
+			if (type === 'node') {
+				this.loadNodeFromFile('CUSTOM', fileName, filePath);
+			} else if (type === 'credentials') {
+				this.loadCredentialFromFile(fileName, filePath);
+			}
+		}
+	}
+}
+
+/**
+ * Loader for source files of nodes and credentials located in a package dir,
+ * e.g. /nodes-base or community packages.
+ */
+export class PackageDirectoryLoader extends DirectoryLoader {
+	packageName = '';
+
+	packageJson!: n8n.PackageJson;
+
+	async readPackageJson() {
+		this.packageJson = await this.readJSON('package.json');
+		this.packageName = this.packageJson.name;
+	}
+
+	override async loadAll() {
+		await this.readPackageJson();
+
+		const { n8n } = this.packageJson;
+		if (!n8n) return;
+
+		const { nodes, credentials } = n8n;
+
+		if (Array.isArray(credentials)) {
+			for (const credential of credentials) {
+				const filePath = this.resolvePath(credential);
+				const [credentialName] = path.parse(credential).name.split('.');
+
+				this.loadCredentialFromFile(credentialName, filePath);
+			}
+		}
+
+		if (Array.isArray(nodes)) {
+			for (const node of nodes) {
+				const filePath = this.resolvePath(node);
+				const [nodeName] = path.parse(node).name.split('.');
+
+				this.loadNodeFromFile(this.packageName, nodeName, filePath);
+			}
+		}
+
+		Logger.info(`Loaded all credentials and nodes from ${this.packageName}`, {
```

Comment:
```suggestion
		Logger.debug(`Loaded all credentials and nodes from ${this.packageName}`, {
```

Reply:
it was `info` originally, but [omar suggested that we use `debug` instead](https://github.com/n8n-io/n8n/pull/4577#discussion_r1024157051). 
I can't decide. What do you think?